pubsub/gossipsub/gossipsub-v1.2.md

```diff
@@ -4,16 +4,103 @@
 |-----------------|---------------------------|--------|-----------------|
 | 1A              | Working Draft             | Active | r1, 2023-07-14  |
 
-Authors:
+Authors: [@Nashatyrev], [@Menduist]
 
-Interest Group: [@vyzo]
+Interest Group: [@vyzo], [@Nashatyrev], [@Menduist]
 
 [@vyzo]: https://github.com/vyzo
+[@Nashatyrev]: https://github.com/Nashatyrev
+[@Menduist]: https://github.com/Menduist
 
 See the [lifecycle document][lifecycle-spec] for context about maturity level and spec status.
 
 [lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
 
-## Overview
+# Overview
+
+This document aims to provide a minimal extension to the [gossipsub
+v1.1](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md)
+protocol.
+
+The proposed extensions are backwards-compatible and aim to enhance the
+efficiency (minimize amplification/duplicates and decrease message latency) of
+the gossip mesh networks for larger messages. 
+
+In more specific terms, a new control message is introduced: `IDONTWANT`. It's primarily 
+intended to notify mesh peers that the node already received a message and there is no 
+need to send its duplicate.
+
+# Specification
+
+## Protocol Id
+
+Nodes that support this Gossipsub extension should additionally advertise the
+version number `1.2.0`. Gossipsub nodes can advertise their own protocol-id
+prefix, by default this is `meshsub` giving the default protocol id:
+- `/meshsub/1.2.0`
+
+## Parameters
+
+This section lists the configuration parameters that needs to agreed on across clients to avoid 
+ peer penalizations
+
+| Parameter                | Description                                                      | Reasonable Default |
+|--------------------------|------------------------------------------------------------------|--------------|
+| `max_idontwant_messages` | The maximum number of `IDONTWANT` messages per heartbeat per peer | ???  |
```

Comment:
I guess this probably needs to be per-topic per heartbeat, rather than a total per heartbeat. 

It seems it could be tied in with the scoring for mesh message delivery rate. I.e the more messages we are expecting per topic, the more IDONTWANT messages we would expect to receive. 

One thought would be to add a behaviour penalty, similar to broken promises, if the number of IDONTWANT messages received from a peer exceeds the mesh message delivery rate. 

We intend to implement this fairly soon. Perhaps we can leave the scoring penalty here for a future PR if we dont want to specify it now.

Reply:
Tracking `IDONTWANT` by topics would be perfect, however you don't know a message topic by its ID unless you already received this message.
`IDONTWANT` message is almost semantically equivalent to `IHAVE` message, so probably should have a similar anti-spam protection mechanism? 
Probably it could make sense to setup reasonable default: `maxIHaveLength * maxIHaveMessages` ?

Reply:
> Is it a good reason to include topics to IDONTWANT and not care what IHAVE and IWANT are doing?

Including a topic would increase `IDONTWANT` traffic 2-3 times (messageID is 20 bytes, topic could be up to around 40 bytes). Not sure if it's worth just to enable more advanced spam protection...

Reply:
Please also consider that `IHAVE` is sent on heartbeat with a batch of messages which may be grouped by topics and the topic overhead here could not be that significant. 
`IDONTWANT` on the other hand is intended to be flushed immediately and would most likely contain just a single messageID so the topic overhead would be more significant

Reply:
We have built a version of this and are going to start testing. It doesn't discriminate based on topic and the implementation seems fine to me. 

Perhaps this first version is topic-independent and if we need to go back and handle topics in a subsequent PR we can do it (with the scoring)

Reply:
Just realized that unlike `IHAVE` the `IDONTWANT` message IDs are never checked and thus could legally contain any random data. The `IDONTWANT` message IDs are supposed to be collected in a set which might be vulnerable to the hash map attack (we already fixed one in the jvm-libp2p impl). 
So it probably makes sense to: 
- Propose a more conservative limit. 
- Add a note to the spec to warn implementers about that possible issue and may be recommend to avoid using HashMap/HashSet implementations with weak hash function (like in Java)

Reply:
> The `IDONTWANT` message IDs are supposed to be collected in a set which might be vulnerable to the hash map attack

I think the implementers tend to use a hash set to remember the IDs that are seen in IDONTWANTs, but it's unreasonable to use the same hash set for `max_idontwant_messages`.

What would be reasonable is to just use a counter and increment it when seeing an IDONTWANT message without using a hash set at all.

Reply:
> What would be reasonable is to just use a counter and increment it when seeing an IDONTWANT message without using a hash set at all.

@ppopth I'm not sure I'm following you. What are we supposed to count? It's expected that message IDs from a signle peer `IDONTWANT` messages would be unique.

Reply:
not sure it's been mentioned, but because the message id:s are not validated, it's possible to send (very) large ones - we're addressing this by hashing the message id which, together with the max-per-heartbeat gives a bound on mem usage overall.

Reply:
> not sure it's been mentioned, but because the message id:s are not validated, it's possible to send (very) large ones - we're addressing this by hashing (and salting) the message id which, together with the max-per-heartbeat gives a bound on mem usage overall.

This is a good point. I will do the same thing in `go-libp2p`.

Reply:
> not cleared at all

we use `mcache_len` for clearing it

Reply:
> we use `mcache_len` for clearing it

Yeah, what I meant is clearing it is the implementation detail and out of scope of the spec.

Reply:
I suspect it might be valuable to specify it - ie effectively, after mcache expires, the message should no longer exist and implementations should be able to rely on them being "resent" if they resurface after that time - this more faithfully keeps the protocol consistent in this aspect

Reply:
>  I don't see why using a weak hash function for unwanted is a problem.

@ppopth I was meaning this [HashDoS attack](https://en.wikipedia.org/wiki/Collision_attack#Hash_flooding). 
The messageIDs in `IDONTWANT` are not validated: neither upon receive nor later (as for `IHAVE`). Thus an adversary may generate and send significant amount of messageIDs which yields the same hashCode in the context of `unwanted` hash set. 
Actually this attack vector could be addressed on the implementation side pretty easily. I just wanted to add a 'warning note' for implementers into the spec

Reply:
> not sure it's been mentioned, but because the message id:s are not validated

@arnetheduck I believe as far as the `messageId` function is application specific, the validation of a `messageId` is left for an application responsibility. I believe Ethereum clients should all have such validation